tally/src/counter_impl.cc

```diff
@@ -0,0 +1,56 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <string>
+
+#include "tally/src/counter_impl.h"
+
+namespace tally {
+
+CounterImpl::CounterImpl() : current_(0), previous_(0) {}
+
+void CounterImpl::Inc() { Inc(1); }
+
+void CounterImpl::Inc(int64_t delta) { current_ += delta; }
+
+void CounterImpl::Report(const std::string &name,
+                         const std::map<std::string, std::string> &tags,
+                         StatsReporter *reporter) {
+  auto val = Value();
+  if (val != 0 && reporter != nullptr) {
+    reporter->ReportCounter(name, tags, val);
+  }
+}
+
+int64_t CounterImpl::Value() {
+  // Load the previous value second so we don't lose any increments.
+  auto current = current_.load();
+  auto previous = previous_.load();
```

Comment:
I think the type-casting operator is overloaded to do the load implicitly. So another option is:

```
const auto current = static_cast<int64_t>(current_);
const auto previous = static_cast<int64_t>(previous_);
```

Reply:
Yep! I just found using `load` to be simpler myself.